File: swift-bridge/Sources/CoreAnimationBridge/Emitter.swift
import CoreGraphics
import Foundation
import QuartzCore

@_cdecl("ca_emitter_layer_new")
public func ca_emitter_layer_new() -> UnsafeMutableRawPointer? {
    caRetain(CAEmitterLayer())
}

@_cdecl("ca_emitter_layer_set_emitter_cells")
public func ca_emitter_layer_set_emitter_cells(_ handle: UnsafeMutableRawPointer?, _ cells: UnsafePointer<UnsafeMutableRawPointer?>?, _ count: Int) {
    guard let layer: CAEmitterLayer = caBorrow(handle) else { return }
    guard let cells, count > 0 else {
        layer.emitterCells = nil
        return
    }
    layer.emitterCells = (0..<count).compactMap { index in
        let handle = cells[index]
        let cell: CAEmitterCell? = caBorrow(handle)
        return cell
    }
}

@_cdecl("ca_emitter_layer_emitter_cell_count")
public func ca_emitter_layer_emitter_cell_count(_ handle: UnsafeMutableRawPointer?) -> Int {
    guard let layer: CAEmitterLayer = caBorrow(handle) else { return 0 }
    return layer.emitterCells?.count ?? 0
}

@_cdecl("ca_emitter_layer_emitter_cell_at")
public func ca_emitter_layer_emitter_cell_at(_ handle: UnsafeMutableRawPointer?, _ index: Int) -> UnsafeMutableRawPointer? {
    guard let layer: CAEmitterLayer = caBorrow(handle), let cells = layer.emitterCells, index >= 0, index < cells.count else { return nil }
    return caRetain(cells[index])
}

@_cdecl("ca_emitter_layer_get_birth_rate")
public func ca_emitter_layer_get_birth_rate(_ handle: UnsafeMutableRawPointer?) -> Float {
    guard let layer: CAEmitterLayer = caBorrow(handle) else { return 0 }
    return layer.birthRate
}

@_cdecl("ca_emitter_layer_set_birth_rate")
public func ca_emitter_layer_set_birth_rate(_ handle: UnsafeMutableRawPointer?, _ value: Float) {
    guard let layer: CAEmitterLayer = caBorrow(handle) else { return }
    layer.birthRate = value
}

@_cdecl("ca_emitter_layer_get_lifetime")
public func ca_emitter_layer_get_lifetime(_ handle: UnsafeMutableRawPointer?) -> Float {
    guard let layer: CAEmitterLayer = caBorrow(handle) else { return 0 }
    return layer.lifetime
}

@_cdecl("ca_emitter_layer_set_lifetime")
public func ca_emitter_layer_set_lifetime(_ handle: UnsafeMutableRawPointer?, _ value: Float) {
    guard let layer: CAEmitterLayer = caBorrow(handle) else { return }
    layer.lifetime = value
}

@_cdecl("ca_emitter_layer_get_emitter_position")
public func ca_emitter_layer_get_emitter_position(_ handle: UnsafeMutableRawPointer?, _ outPoint: UnsafeMutableRawPointer?) -> Bool {
    guard let layer: CAEmitterLayer = caBorrow(handle) else { return false }
    return caWritePoint(layer.emitterPosition, out: outPoint)
}

@_cdecl("ca_emitter_layer_set_emitter_position")
public func ca_emitter_layer_set_emitter_position(_ handle: UnsafeMutableRawPointer?, _ x: Double, _ y: Double) {
    guard let layer: CAEmitterLayer = caBorrow(handle) else { return }
    layer.emitterPosition = CGPoint(x: x, y: y)
}

@_cdecl("ca_emitter_layer_get_emitter_size")
public func ca_emitter_layer_get_emitter_size(_ handle: UnsafeMutableRawPointer?, _ outSize: UnsafeMutableRawPointer?) -> Bool {
    guard let layer: CAEmitterLayer = caBorrow(handle) else { return false }
    return caWriteSize(layer.emitterSize, out: outSize)
}

@_cdecl("ca_emitter_layer_set_emitter_size")
public func ca_emitter_layer_set_emitter_size(_ handle: UnsafeMutableRawPointer?, _ width: Double, _ height: Double) {
    guard let layer: CAEmitterLayer = caBorrow(handle) else { return }
    layer.emitterSize = CGSize(width: width, height: height)
}

@_cdecl("ca_emitter_layer_get_emitter_shape")
public func ca_emitter_layer_get_emitter_shape(_ handle: UnsafeMutableRawPointer?) -> Int32 {
    guard let layer: CAEmitterLayer = caBorrow(handle) else { return 0 }
    return caEmitterShapeRaw(layer.emitterShape)
}

@_cdecl("ca_emitter_layer_set_emitter_shape")
public func ca_emitter_layer_set_emitter_shape(_ handle: UnsafeMutableRawPointer?, _ value: Int32) {
    guard let layer: CAEmitterLayer = caBorrow(handle) else { return }
    layer.emitterShape = caEmitterShape(value)
}

@_cdecl("ca_emitter_layer_get_emitter_mode")
public func ca_emitter_layer_get_emitter_mode(_ handle: UnsafeMutableRawPointer?) -> Int32 {
    guard let layer: CAEmitterLayer = caBorrow(handle) else { return 3 }
    return caEmitterModeRaw(layer.emitterMode)
}

@_cdecl("ca_emitter_layer_set_emitter_mode")
public func ca_emitter_layer_set_emitter_mode(_ handle: UnsafeMutableRawPointer?, _ value: Int32) {
    guard let layer: CAEmitterLayer = caBorrow(handle) else { return }
    layer.emitterMode = caEmitterMode(value)
}

@_cdecl("ca_emitter_layer_get_render_mode")
public func ca_emitter_layer_get_render_mode(_ handle: UnsafeMutableRawPointer?) -> Int32 {
    guard let layer: CAEmitterLayer = caBorrow(handle) else { return 0 }
    return caEmitterRenderModeRaw(layer.renderMode)
}

@_cdecl("ca_emitter_layer_set_render_mode")
public func ca_emitter_layer_set_render_mode(_ handle: UnsafeMutableRawPointer?, _ value: Int32) {
    guard let layer: CAEmitterLayer = caBorrow(handle) else { return }
    layer.renderMode = caEmitterRenderMode(value)
}

@_cdecl("ca_emitter_layer_get_velocity")
public func ca_emitter_layer_get_velocity(_ handle: UnsafeMutableRawPointer?) -> Float {
    guard let layer: CAEmitterLayer = caBorrow(handle) else { return 0 }
    return layer.velocity
}

@_cdecl("ca_emitter_layer_set_velocity")
public func ca_emitter_layer_set_velocity(_ handle: UnsafeMutableRawPointer?, _ value: Float) {
    guard let layer: CAEmitterLayer = caBorrow(handle) else { return }
    layer.velocity = value
}

@_cdecl("ca_emitter_layer_get_scale")
public func ca_emitter_layer_get_scale(_ handle: UnsafeMutableRawPointer?) -> Float {
    guard let layer: CAEmitterLayer = caBorrow(handle) else { return 0 }
    return layer.scale
}

@_cdecl("ca_emitter_layer_set_scale")
public func ca_emitter_layer_set_scale(_ handle: UnsafeMutableRawPointer?, _ value: Float) {
    guard let layer: CAEmitterLayer = caBorrow(handle) else { return }
    layer.scale = value
}

@_cdecl("ca_emitter_cell_new")
public func ca_emitter_cell_new() -> UnsafeMutableRawPointer? {
    caRetain(CAEmitterCell())
}

@_cdecl("ca_emitter_cell_set_name")
public func ca_emitter_cell_set_name(_ handle: UnsafeMutableRawPointer?, _ value: UnsafePointer<CChar>?) {
    guard let cell: CAEmitterCell = caBorrow(handle) else { return }
    cell.name = caCString(value)
}

@_cdecl("ca_emitter_cell_get_name")
public func ca_emitter_cell_get_name(_ handle: UnsafeMutableRawPointer?) -> UnsafeMutablePointer<CChar>? {
    guard let cell: CAEmitterCell = caBorrow(handle) else { return nil }
    return caDup(cell.name)
}

@_cdecl("ca_emitter_cell_get_enabled")
public func ca_emitter_cell_get_enabled(_ handle: UnsafeMutableRawPointer?) -> Bool {
    guard let cell: CAEmitterCell = caBorrow(handle) else { return false }
    return cell.isEnabled
}

@_cdecl("ca_emitter_cell_set_enabled")
public func ca_emitter_cell_set_enabled(_ handle: UnsafeMutableRawPointer?, _ value: Bool) {
    guard let cell: CAEmitterCell = caBorrow(handle) else { return }
    cell.isEnabled = value
}

@_cdecl("ca_emitter_cell_get_birth_rate")
public func ca_emitter_cell_get_birth_rate(_ handle: UnsafeMutableRawPointer?) -> Float {
    guard let cell: CAEmitterCell = caBorrow(handle) else { return 0 }
    return cell.birthRate
}

@_cdecl("ca_emitter_cell_set_birth_rate")
public func ca_emitter_cell_set_birth_rate(_ handle: UnsafeMutableRawPointer?, _ value: Float) {
    guard let cell: CAEmitterCell = caBorrow(handle) else { return }
    cell.birthRate = value
}

@_cdecl("ca_emitter_cell_get_lifetime")
public func ca_emitter_cell_get_lifetime(_ handle: UnsafeMutableRawPointer?) -> Float {
    guard let cell: CAEmitterCell = caBorrow(handle) else { return 0 }
    return cell.lifetime
}

@_cdecl("ca_emitter_cell_set_lifetime")
public func ca_emitter_cell_set_lifetime(_ handle: UnsafeMutableRawPointer?, _ value: Float) {
    guard let cell: CAEmitterCell = caBorrow(handle) else { return }
    cell.lifetime = value
}

@_cdecl("ca_emitter_cell_get_velocity")
public func ca_emitter_cell_get_velocity(_ handle: UnsafeMutableRawPointer?) -> Double {
    guard let cell: CAEmitterCell = caBorrow(handle) else { return 0 }
    return cell.velocity
}

@_cdecl("ca_emitter_cell_set_velocity")
public func ca_emitter_cell_set_velocity(_ handle: UnsafeMutableRawPointer?, _ value: Double) {
    guard let cell: CAEmitterCell = caBorrow(handle) else { return }
    cell.velocity = value
}

@_cdecl("ca_emitter_cell_get_scale")
public func ca_emitter_cell_get_scale(_ handle: UnsafeMutableRawPointer?) -> Double {
    guard let cell: CAEmitterCell = caBorrow(handle) else { return 0 }
    return cell.scale
}

@_cdecl("ca_emitter_cell_set_scale")
public func ca_emitter_cell_set_scale(_ handle: UnsafeMutableRawPointer?, _ value: Double) {
    guard let cell: CAEmitterCell = caBorrow(handle) else { return }
    cell.scale = value
}

@_cdecl("ca_emitter_cell_get_emission_range")
public func ca_emitter_cell_get_emission_range(_ handle: UnsafeMutableRawPointer?) -> Double {
    guard let cell: CAEmitterCell = caBorrow(handle) else { return 0 }
    return cell.emissionRange
}

@_cdecl("ca_emitter_cell_set_emission_range")
public func ca_emitter_cell_set_emission_range(_ handle: UnsafeMutableRawPointer?, _ value: Double) {
    guard let cell: CAEmitterCell = caBorrow(handle) else { return }
    cell.emissionRange = value
}

@_cdecl("ca_emitter_cell_get_emission_longitude")
public func ca_emitter_cell_get_emission_longitude(_ handle: UnsafeMutableRawPointer?) -> Double {
    guard let cell: CAEmitterCell = caBorrow(handle) else { return 0 }
    return cell.emissionLongitude
}

@_cdecl("ca_emitter_cell_set_emission_longitude")
public func ca_emitter_cell_set_emission_longitude(_ handle: UnsafeMutableRawPointer?, _ value: Double) {
    guard let cell: CAEmitterCell = caBorrow(handle) else { return }
    cell.emissionLongitude = value
}

@_cdecl("ca_emitter_cell_set_color")
public func ca_emitter_cell_set_color(_ handle: UnsafeMutableRawPointer?, _ colorHandle: UnsafeMutableRawPointer?) {
    guard let cell: CAEmitterCell = caBorrow(handle) else { return }
    let color: CGColor? = caBorrow(colorHandle)
    cell.color = color
}

@_cdecl("ca_emitter_cell_get_color")
public func ca_emitter_cell_get_color(_ handle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let cell: CAEmitterCell = caBorrow(handle), let color = cell.color else { return nil }
    return caRetain(color)
}

@_cdecl("ca_emitter_cell_set_contents")
public func ca_emitter_cell_set_contents(_ handle: UnsafeMutableRawPointer?, _ imageHandle: UnsafeMutableRawPointer?) {
    guard let cell: CAEmitterCell = caBorrow(handle) else { return }
    let image: CGImage? = caBorrow(imageHandle)
    cell.contents = image
}

@_cdecl("ca_emitter_cell_get_contents")
public func ca_emitter_cell_get_contents(_ handle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let cell: CAEmitterCell = caBorrow(handle), let contents = cell.contents else { return nil }
    return caRetain(contents as AnyObject)
}

@_cdecl("ca_emitter_cell_get_alpha_speed")
public func ca_emitter_cell_get_alpha_speed(_ handle: UnsafeMutableRawPointer?) -> Float {
    guard let cell: CAEmitterCell = caBorrow(handle) else { return 0 }
    return cell.alphaSpeed
}

@_cdecl("ca_emitter_cell_set_alpha_speed")
public func ca_emitter_cell_set_alpha_speed(_ handle: UnsafeMutableRawPointer?, _ value: Float) {
    guard let cell: CAEmitterCell = caBorrow(handle) else { return }
    cell.alphaSpeed = value
}

File: src/emitter.rs
use std::ffi::CStr;
use std::ops::Deref;

use apple_cf::cg::CGImage;
use apple_cf::cg::{CGPoint, CGSize};

use crate::color::Color;
use crate::layer::{Layer, LayerLike};
use crate::private::{cstring_from_str, handle_type};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(i32)]
pub enum EmitterShape {
    Point = 0,
    Line = 1,
    Rectangle = 2,
    Cuboid = 3,
    Circle = 4,
    Sphere = 5,
}

impl EmitterShape {
    const fn from_raw(value: i32) -> Self {
        match value {
            1 => Self::Line,
            2 => Self::Rectangle,
            3 => Self::Cuboid,
            4 => Self::Circle,
            5 => Self::Sphere,
            _ => Self::Point,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(i32)]
pub enum EmitterMode {
    Points = 0,
    Outline = 1,
    Surface = 2,
    Volume = 3,
}

impl EmitterMode {
    const fn from_raw(value: i32) -> Self {
        match value {
            1 => Self::Outline,
            2 => Self::Surface,
            _ => Self::Volume,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(i32)]
pub enum EmitterRenderMode {
    Unordered = 0,
    OldestFirst = 1,
    OldestLast = 2,
    BackToFront = 3,
    Additive = 4,
}

impl EmitterRenderMode {
    const fn from_raw(value: i32) -> Self {
        match value {
            1 => Self::OldestFirst,
            2 => Self::OldestLast,
            3 => Self::BackToFront,
            4 => Self::Additive,
            _ => Self::Unordered,
        }
    }
}

#[derive(Debug, Clone)]
pub struct EmitterLayer {
    inner: Layer,
}

impl EmitterLayer {
    #[must_use]
    pub fn new() -> Option<Self> {
        unsafe { Layer::from_raw(crate::ffi::ca_emitter_layer_new()) }.map(|inner| Self { inner })
    }

    pub fn set_emitter_cells(&self, cells: &[&EmitterCell]) {
        let raw: Vec<*mut core::ffi::c_void> = cells.iter().map(|cell| cell.as_ptr()).collect();
        unsafe {
            crate::ffi::ca_emitter_layer_set_emitter_cells(
                self.as_layer_ptr(),
                raw.as_ptr(),
                raw.len(),
            )
        };
    }

    #[must_use]
    pub fn emitter_cells(&self) -> Vec<EmitterCell> {
        let count = unsafe { crate::ffi::ca_emitter_layer_emitter_cell_count(self.as_layer_ptr()) };
        (0..count)
            .filter_map(|index| unsafe {
                EmitterCell::from_raw(crate::ffi::ca_emitter_layer_emitter_cell_at(
                    self.as_layer_ptr(),
                    index,
                ))
            })
            .collect()
    }

    #[must_use]
    pub fn birth_rate(&self) -> f32 {
        unsafe { crate::ffi::ca_emitter_layer_get_birth_rate(self.as_layer_ptr()) }
    }

    pub fn set_birth_rate(&self, value: f32) {
        unsafe { crate::ffi::ca_emitter_layer_set_birth_rate(self.as_layer_ptr(), value) };
    }

    #[must_use]
    pub fn lifetime(&self) -> f32 {
        unsafe { crate::ffi::ca_emitter_layer_get_lifetime(self.as_layer_ptr()) }
    }

    pub fn set_lifetime(&self, value: f32) {
        unsafe { crate::ffi::ca_emitter_layer_set_lifetime(self.as_layer_ptr(), value) };
    }

    #[must_use]
    pub fn emitter_position(&self) -> CGPoint {
        let mut point = CGPoint::zero();
        let ok = unsafe {
            crate::ffi::ca_emitter_layer_get_emitter_position(
                self.as_layer_ptr(),
                (&mut point as *mut CGPoint).cast::<core::ffi::c_void>(),
            )
        };
        if ok {
            point
        } else {
            CGPoint::zero()
        }
    }

    pub fn set_emitter_position(&self, value: CGPoint) {
        unsafe {
            crate::ffi::ca_emitter_layer_set_emitter_position(self.as_layer_ptr(), value.x, value.y)
        };
    }

    #[must_use]
    pub fn emitter_size(&self) -> CGSize {
        let mut size = CGSize::zero();
        let ok = unsafe {
            crate::ffi::ca_emitter_layer_get_emitter_size(
                self.as_layer_ptr(),
                (&mut size as *mut CGSize).cast::<core::ffi::c_void>(),
            )
        };
        if ok {
            size
        } else {
            CGSize::zero()
        }
    }

    pub fn set_emitter_size(&self, value: CGSize) {
        unsafe {
            crate::ffi::ca_emitter_layer_set_emitter_size(
                self.as_layer_ptr(),
                value.width,
                value.height,
            )
        };
    }

    #[must_use]
    pub fn emitter_shape(&self) -> EmitterShape {
        EmitterShape::from_raw(unsafe {
            crate::ffi::ca_emitter_layer_get_emitter_shape(self.as_layer_ptr())
        })
    }

    pub fn set_emitter_shape(&self, value: EmitterShape) {
        unsafe {
            crate::ffi::ca_emitter_layer_set_emitter_shape(self.as_layer_ptr(), value as i32)
        };
    }

    #[must_use]
    pub fn emitter_mode(&self) -> EmitterMode {
        EmitterMode::from_raw(unsafe {
            crate::ffi::ca_emitter_layer_get_emitter_mode(self.as_layer_ptr())
        })
    }

    pub fn set_emitter_mode(&self, value: EmitterMode) {
        unsafe { crate::ffi::ca_emitter_layer_set_emitter_mode(self.as_layer_ptr(), value as i32) };
    }

    #[must_use]
    pub fn render_mode(&self) -> EmitterRenderMode {
        EmitterRenderMode::from_raw(unsafe {
            crate::ffi::ca_emitter_layer_get_render_mode(self.as_layer_ptr())
        })
    }

    pub fn set_render_mode(&self, value: EmitterRenderMode) {
        unsafe { crate::ffi::ca_emitter_layer_set_render_mode(self.as_layer_ptr(), value as i32) };
    }

    #[must_use]
    pub fn velocity(&self) -> f32 {
        unsafe { crate::ffi::ca_emitter_layer_get_velocity(self.as_layer_ptr()) }
    }

    pub fn set_velocity(&self, value: f32) {
        unsafe { crate::ffi::ca_emitter_layer_set_velocity(self.as_layer_ptr(), value) };
    }

    #[must_use]
    pub fn scale(&self) -> f32 {
        unsafe { crate::ffi::ca_emitter_layer_get_scale(self.as_layer_ptr()) }
    }

    pub fn set_scale(&self, value: f32) {
        unsafe { crate::ffi::ca_emitter_layer_set_scale(self.as_layer_ptr(), value) };
    }
}

impl Deref for EmitterLayer {
    type Target = Layer;

    fn deref(&self) -> &Self::Target {
        &self.inner
    }
}

impl LayerLike for EmitterLayer {
    fn as_layer_ptr(&self) -> *mut core::ffi::c_void {
        self.inner.as_ptr()
    }
}

handle_type!(EmitterCell);

impl EmitterCell {
    #[must_use]
    pub fn new() -> Option<Self> {
        unsafe { Self::from_raw(crate::ffi::ca_emitter_cell_new()) }
    }

    pub fn set_name(&self, value: &str) {
        if let Some(value) = cstring_from_str(value) {
            unsafe { crate::ffi::ca_emitter_cell_set_name(self.as_ptr(), value.as_ptr()) };
        }
    }

    #[must_use]
    pub fn name(&self) -> Option<String> {
        take_c_string(unsafe { crate::ffi::ca_emitter_cell_get_name(self.as_ptr()) })
    }

    #[must_use]
    pub fn enabled(&self) -> bool {
        unsafe { crate::ffi::ca_emitter_cell_get_enabled(self.as_ptr()) }
    }

    pub fn set_enabled(&self, value: bool) {
        unsafe { crate::ffi::ca_emitter_cell_set_enabled(self.as_ptr(), value) };
    }

    #[must_use]
    pub fn birth_rate(&self) -> f32 {
        unsafe { crate::ffi::ca_emitter_cell_get_birth_rate(self.as_ptr()) }
    }

    pub fn set_birth_rate(&self, value: f32) {
        unsafe { crate::ffi::ca_emitter_cell_set_birth_rate(self.as_ptr(), value) };
    }

    #[must_use]
    pub fn lifetime(&self) -> f32 {
        unsafe { crate::ffi::ca_emitter_cell_get_lifetime(self.as_ptr()) }
    }

    pub fn set_lifetime(&self, value: f32) {
        unsafe { crate::ffi::ca_emitter_cell_set_lifetime(self.as_ptr(), value) };
    }

    #[must_use]
    pub fn velocity(&self) -> f64 {
        unsafe { crate::ffi::ca_emitter_cell_get_velocity(self.as_ptr()) }
    }

    pub fn set_velocity(&self, value: f64) {
        unsafe { crate::ffi::ca_emitter_cell_set_velocity(self.as_ptr(), value) };
    }

    #[must_use]
    pub fn scale(&self) -> f64 {
        unsafe { crate::ffi::ca_emitter_cell_get_scale(self.as_ptr()) }
    }

    pub fn set_scale(&self, value: f64) {
        unsafe { crate::ffi::ca_emitter_cell_set_scale(self.as_ptr(), value) };
    }

    #[must_use]
    pub fn emission_range(&self) -> f64 {
        unsafe { crate::ffi::ca_emitter_cell_get_emission_range(self.as_ptr()) }
    }

    pub fn set_emission_range(&self, value: f64) {
        unsafe { crate::ffi::ca_emitter_cell_set_emission_range(self.as_ptr(), value) };
    }

    #[must_use]
    pub fn emission_longitude(&self) -> f64 {
        unsafe { crate::ffi::ca_emitter_cell_get_emission_longitude(self.as_ptr()) }
    }

    pub fn set_emission_longitude(&self, value: f64) {
        unsafe { crate::ffi::ca_emitter_cell_set_emission_longitude(self.as_ptr(), value) };
    }

    pub fn set_color(&self, color: Option<&Color>) {
        unsafe {
            crate::ffi::ca_emitter_cell_set_color(
                self.as_ptr(),
                color.map_or(core::ptr::null_mut(), Color::as_ptr),
            )
        };
    }

    #[must_use]
    pub fn color(&self) -> Option<Color> {
        unsafe { Color::from_raw(crate::ffi::ca_emitter_cell_get_color(self.as_ptr())) }
    }

    pub fn set_contents(&self, image: Option<&CGImage>) {
        unsafe {
            crate::ffi::ca_emitter_cell_set_contents(
                self.as_ptr(),
                image.map_or(core::ptr::null_mut(), CGImage::as_ptr),
            )
        };
    }

    #[must_use]
    pub fn contents(&self) -> Option<CGImage> {
        let ptr = unsafe { crate::ffi::ca_emitter_cell_get_contents(self.as_ptr()) };
        if ptr.is_null() {
            None
        } else {
            Some(unsafe { CGImage::from_raw(ptr) })
        }
    }

    #[must_use]
    pub fn alpha_speed(&self) -> f32 {
        unsafe { crate::ffi::ca_emitter_cell_get_alpha_speed(self.as_ptr()) }
    }

    pub fn set_alpha_speed(&self, value: f32) {
        unsafe { crate::ffi::ca_emitter_cell_set_alpha_speed(self.as_ptr(), value) };
    }
}

fn take_c_string(ptr: *mut libc::c_char) -> Option<String> {
    if ptr.is_null() {
        return None;
    }
    let value = unsafe { CStr::from_ptr(ptr) }
        .to_string_lossy()
        .into_owned();
    unsafe { libc::free(ptr.cast()) };
    Some(value)
}
